// tslint-to-eslint-config/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.joinConfigConversionResults = exports.formatOutput = exports.findReportedConfiguration = exports.findOriginalConfigurations = exports.createESLintConfiguration = exports.convertTSLintConfig = exports.convertFileComments = void 0;
var convertFileCommentsStandalone_1 = require("./api/convertFileCommentsStandalone");
Object.defineProperty(exports, "convertFileComments", { enumerable: true, get: function () { return convertFileCommentsStandalone_1.convertFileCommentsStandalone; } });
var convertTSLintConfigStandalone_1 = require("./api/convertTSLintConfigStandalone");
Object.defineProperty(exports, "convertTSLintConfig", { enumerable: true, get: function () { return convertTSLintConfigStandalone_1.convertTSLintConfigStandalone; } });
var createESLintConfigurationStandalone_1 = require("./api/createESLintConfigurationStandalone");
Object.defineProperty(exports, "createESLintConfiguration", { enumerable: true, get: function () { return createESLintConfigurationStandalone_1.createESLintConfigurationStandalone; } });
var findOriginalConfigurationsStandalone_1 = require("./api/findOriginalConfigurationsStandalone");
Object.defineProperty(exports, "findOriginalConfigurations", { enumerable: true, get: function () { return findOriginalConfigurationsStandalone_1.findOriginalConfigurationsStandalone; } });
var findReportedConfigurationStandalone_1 = require("./api/findReportedConfigurationStandalone");
Object.defineProperty(exports, "findReportedConfiguration", { enumerable: true, get: function () { return findReportedConfigurationStandalone_1.findReportedConfigurationStandalone; } });
var formatOutput_1 = require("./converters/lintConfigs/formatting/formatOutput");
Object.defineProperty(exports, "formatOutput", { enumerable: true, get: function () { return formatOutput_1.formatOutput; } });
var joinConfigConversionResults_1 = require("./converters/lintConfigs/joinConfigConversionResults");
Object.defineProperty(exports, "joinConfigConversionResults", { enumerable: true, get: function () { return joinConfigConversionResults_1.joinConfigConversionResults; } });
__exportStar(require("./converters/lintConfigs/rules/types"), exports);
__exportStar(require("./types"), exports);
