// tslint-to-eslint-config/src/converters/lintConfigs/rules/ruleConverters/no-magic-numbers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertNoMagicNumbers = void 0;
const convertNoMagicNumbers = (tslintRule) => {
    return {
        rules: [
            {
                ...collectArgumentsAndNotices(tslintRule.ruleArguments),
                ruleName: "no-magic-numbers",
            },
        ],
    };
};
exports.convertNoMagicNumbers = convertNoMagicNumbers;
const collectArgumentsAndNotices = (ruleArguments) => {
    if (ruleArguments.length === 0) {
        return undefined;
    }
    if (ruleArguments.length !== 1 || typeof ruleArguments[0] === "number") {
        return {
            ruleArguments: [
                {
                    ignore: ruleArguments,
                },
            ],
        };
    }
    return {
        ...("ignore-jsx" in ruleArguments[0] && {
            notices: ["JSX syntax will no longer be ignored."],
        }),
        ...("allowed-numbers" in ruleArguments[0] && {
            ruleArguments: [{ ignore: ruleArguments[0]["allowed-numbers"] }],
        }),
    };
};
