// @ngrx/store/esm2015/testing/mock_reducer_manager.js
/**
 * @fileoverview added by tsickle
 * Generated from: mock_reducer_manager.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
export class MockReducerManager extends BehaviorSubject {
    constructor() {
        super((/**
         * @return {?}
         */
        () => undefined));
    }
    /**
     * @param {?} feature
     * @return {?}
     */
    addFeature(feature) {
        /* noop */
    }
    /**
     * @param {?} feature
     * @return {?}
     */
    addFeatures(feature) {
        /* noop */
    }
    /**
     * @param {?} feature
     * @return {?}
     */
    removeFeature(feature) {
        /* noop */
    }
    /**
     * @param {?} features
     * @return {?}
     */
    removeFeatures(features) {
        /* noop */
    }
    /**
     * @param {?} key
     * @param {?} reducer
     * @return {?}
     */
    addReducer(key, reducer) {
        /* noop */
    }
    /**
     * @param {?} reducers
     * @return {?}
     */
    addReducers(reducers) {
        /* noop */
    }
    /**
     * @param {?} featureKey
     * @return {?}
     */
    removeReducer(featureKey) {
        /* noop */
    }
    /**
     * @param {?} featureKeys
     * @return {?}
     */
    removeReducers(featureKeys) {
        /* noop */
    }
}
MockReducerManager.ɵfac = function MockReducerManager_Factory(t) { return new (t || MockReducerManager)(); };
MockReducerManager.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: MockReducerManager, factory: MockReducerManager.ɵfac });
/** @nocollapse */
MockReducerManager.ctorParameters = () => [];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MockReducerManager, [{
        type: Injectable
    }], function () { return []; }, null); })();
