// @crystalui/angular-lightbox/esm2015/lib/lightbox.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LightboxCommonComponent } from './lightbox-common.component';
import { LightboxComponent } from './lightbox.component';
import { CrystalLightbox } from './lightbox.service';
import { EventService } from './event.service';
import { LightboxDirective } from './lightbox.directive';
import { LightboxGroupDirective } from './lightbox-group.directive';
import * as ɵngcc0 from '@angular/core';
let CrystalLightboxModule = class CrystalLightboxModule {
};
CrystalLightboxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CrystalLightboxModule });
CrystalLightboxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CrystalLightboxModule_Factory(t) { return new (t || CrystalLightboxModule)(); }, providers: [
        CrystalLightbox,
        EventService
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CrystalLightboxModule, { declarations: function () { return [LightboxCommonComponent,
        LightboxComponent,
        LightboxDirective,
        LightboxGroupDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [LightboxDirective,
        LightboxGroupDirective]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CrystalLightboxModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    LightboxCommonComponent,
                    LightboxComponent,
                    LightboxDirective,
                    LightboxGroupDirective
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    LightboxDirective,
                    LightboxGroupDirective
                ],
                providers: [
                    CrystalLightbox,
                    EventService
                ],
                bootstrap: [],
                entryComponents: [
                    LightboxComponent
                ]
            }]
    }], null, null); })();
export { CrystalLightboxModule };
