// tslint-to-eslint-config/src/converters/lintConfigs/joinConfigConversionResults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.joinConfigConversionResults = void 0;
const utils_1 = require("../../utils");
const createEnv_1 = require("./eslint/createEnv");
const formatConvertedRules_1 = require("./formatConvertedRules");
/**
 * Turns a raw ESLint configuration summary into ESLint's configuration shape.
 */
const joinConfigConversionResults = (summarizedResults, originalConfigurations) => {
    const plugins = new Set([...summarizedResults.plugins, "@typescript-eslint"]);
    const { eslint, tslint } = originalConfigurations;
    if (summarizedResults.missing.length !== 0) {
        plugins.add("@typescript-eslint/tslint");
    }
    return utils_1.removeEmptyMembers({
        ...eslint === null || eslint === void 0 ? void 0 : eslint.full,
        env: createEnv_1.createEnv(originalConfigurations),
        ...(eslint && { globals: eslint.raw.globals }),
        ...(summarizedResults.extends && { extends: summarizedResults.extends }),
        parser: "@typescript-eslint/parser",
        parserOptions: {
            project: "tsconfig.json",
            sourceType: "module",
        },
        plugins: Array.from(plugins),
        rules: formatConvertedRules_1.formatConvertedRules(summarizedResults, tslint.full),
    });
};
exports.joinConfigConversionResults = joinConfigConversionResults;
